feat(c/driver_manager,rust/driver_manager): handle virtual environments in driver manager

zeroshade opened the pull request:
Updates the driver manager logic to handle virtual environments used by venv and conda.

venv will set a `VIRTUAL_ENV` environment variable while conda will set `CONDA_PREFIX`. If these environment variables are set, then they are added to the search list of `get_search_paths` / `GetSearchPaths` when the `LOAD_FLAG_SEARCH_ENV` flag is set.

Also updates the docs to mention this.

ianmcook commented on docs/source/format/driver_manifests.rst:
```diff
@@ -431,9 +431,14 @@ Unix-like Platforms
    * ``ADBC_CONFIG_PATH`` is a colon-separated list of directories to search for ``${name}.toml``
+   * ``VIRTUAL_ENV`` is used by `venv <https://docs.python.org/3/library/venv.html>`__ and if set
+     ``$VIRTUAL_ENV/etc/adbc`` will be searched for ``${name}.toml``
+   * ``CONDA_PREFIX`` is used by `conda <https://anaconda.org/anaconda/conda>`__ when using a ``conda``
+     environment (by using ``conda activate env-name``), and if set then ``$CONDA_PREFIX/etc/adbc``
+     will be searched for ``${name}.toml``
```
I think it'd be OK to leave out the details to make this shorter:
```suggestion
   * ``CONDA_PREFIX`` is used by `conda <https://anaconda.org/anaconda/conda>`__ and if set then
     ``$CONDA_PREFIX/etc/adbc`` will be searched for ``${name}.toml``
```

ianmcook commented on docs/source/format/driver_manifests.rst:
```diff
@@ -452,9 +456,15 @@ Windows
    * ``ADBC_CONFIG_PATH`` is a semicolon-separated list of directories to search for ``${name}.toml``
+   * ``VIRTUAL_ENV`` is used by `venv <https://docs.python.org/3/library/venv.html>`__ and if set
+     ``$VIRTUAL_ENV/etc/adbc`` will be searched for ``${name}.toml``
+   * ``CONDA_PREFIX`` is used by `conda <https://anaconda.org/anaconda/conda>`__ when using a ``conda``
+     environment (by using ``conda activate env-name``), and if set then ``$CONDA_PREFIX/etc/adbc``
+     will be searched for ``${name}.toml``
```
Same here
```suggestion
   * ``CONDA_PREFIX`` is used by `conda <https://anaconda.org/anaconda/conda>`__ and if set
     ``$CONDA_PREFIX/etc/adbc`` will be searched for ``${name}.toml``
```

ianmcook commented on docs/source/format/driver_manifests.rst:
```diff
@@ -431,9 +431,13 @@ Unix-like Platforms
    * ``ADBC_CONFIG_PATH`` is a colon-separated list of directories to search for ``${name}.toml``
+   * ``VIRTUAL_ENV`` is used by `venv <https://docs.python.org/3/library/venv.html>`__ and if set
+     ``$VIRTUAL_ENV/etc/adbc`` will be searched for ``${name}.toml``
+   * ``CONDA_PREFIX`` is used by `conda <https://anaconda.org/anaconda/conda>`__ and if set then
```
tiniest nit
```suggestion
   * ``CONDA_PREFIX`` is used by `conda <https://anaconda.org/anaconda/conda>`__ and if set
```

ianmcook commented:
@paleolimbot Thanks for calling this out. I think what's proposed here would be problematic as the default search behavior for driver shared libraries. What makes it reasonable (at least in my thinking) is that it's not that; it's the search behavior for TOML manifests. The TOML manifests are fundamentally an optional affordance for portability and user convenience. I expect that the vast majority of developers building apps that use the driver managers will skip using manifests and instead have the driver managers search directly for driver shared libraries. With all things concerning the manifests, we're operating within a small corner of the overall space, and in this small corner the objective is to afford portability and convenience. In that context, this proposed behavior seems well aligned with that goal.

ianmcook commented:
Ah I see what you mean. Another way to implement this would be to add a config function in the C driver manager that allows wrappers to add paths to the manifest search paths at runtime. Then for example the Python driver manager could check if `VIRTUAL_ENV` is set and add it to the search paths if it is.

But conda is language-agnostic so I think there's a stronger case for keeping `CONDA_PREFIX` in the C driver manager.

zeroshade commented:
@paleolimbot How's the updated version? 

I've put the checks for `CONDA_PREFIX` behind the existence of the `CONDA_BUILD` env var at compile time (since it seems all conda build utilities will set that env var during build) in both the C and Rust driver managers. 

I've then added the `additional_search_path_list` argument to `AdbcFindLoadDriver` and added an appropriate setter for `AdbcDatabase` to manipulate it. Finally, I updated the python driver manager to check if it's in a virtual environment, and if it is it will add the virtual environment prefix +'/etc/adbc' to the search path list. (And updated the R implementation accordingly).

Thoughts?

@lidavidm @ianmcook @amoeba Did I cover everything?

zeroshade commented:
@amoeba correct, that's the current logic. Installing the PyPI adbc_driver_manager package would not perform automatic checking of your conda env, (but it would recognize a venv), and installing it from conda would then recognize the conda_prefix env var and check for it.

I was following @paleolimbot's suggestion of only checking `CONDA_PREFIX` if built for conda. I'm on the fence personally as to which makes more sense.... I'll leave that to you guys to duke out and I'll just implement whatever is decided.

@ianmcook you're right, I didn't update the docs to fit the new behavior yet. Thanks for taking on the updates there. you should be able to just push to this PR any doc updates you make.

lidavidm commented:
> For example, if I install the PyPI adbc_driver_manager package into my Conda environment, driver loading won't search my conda environment by default? It seems cleaner to me to just put the conda stuff alongside the venv stuff in the Python package.

Under Conda you can't assume you'll only be used by Python, so relying entirely on the Python-side code won't work IMO

ianmcook commented:
Reconciling the comments above: How about we:
- Search in `CONDA_PREFIX` when the driver manager was built by conda (which this already does).
- Also add `CONDA_PREFIX` to the search paths in the Python driver manager (coming after `VIRTUAL_ENV` in precedence) regardless of whether it was built by conda.

Does that work for everyone?

lidavidm commented on docs/source/format/driver_manifests.rst:
```diff
@@ -351,7 +351,8 @@ to control which directories will be searched for manifests, with the behavior b
        The type :c:type:`AdbcLoadFlags` is a set of bitflags to control the directories to be searched. The flags are
 
-       * :c:macro:`ADBC_LOAD_FLAG_SEARCH_ENV` - search the environment variable ``ADBC_CONFIG_PATH``
+       * :c:macro:`ADBC_LOAD_FLAG_SEARCH_ENV` - search the directory paths in the environment variable
+       ``ADBC_CONFIG_PATH`` and (when built or installed with conda) search in the conda environment
```
```suggestion
       * :c:macro:`ADBC_LOAD_FLAG_SEARCH_ENV` - search the directory paths in the environment variable
         ``ADBC_CONFIG_PATH`` and (when built or installed with conda) search in the conda environment
```

I think these need to be indended for reST to recognize it as part of the same list

ianmcook replied:
Fixed in e32e72a

lidavidm commented:
Also pre-commit indicates adbc_driver_manager.h needs to be synced to the go/ tree.

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -446,14 +489,22 @@ struct ManagedLibrary {
       return ADBC_STATUS_INVALID_ARGUMENT;
     }
 
+    {
+      auto status = SearchPaths(driver_path, additional_search_paths, info, error);
+      if (status == ADBC_STATUS_OK) {
+        return status;
+      }
+    }
+
```
Currently the `$CONDA_PREFIX/etc/adbc` path (if it exists) will be listed as part of the `ENV` search paths, so it's difficult to insert the user-added paths before that but after the `ADBC_CONFIG_PATH` paths without a refactor. I'll see what I can do, but it's gonna be weird....

ianmcook commented:
@zeroshade can you please comment the `#endif`s like `#endif // _WIN32` to make the conditional logic easier to follow?

ianmcook commented:
In the parts of the code added/changed here, we're careful to use `wchar_t` for env vars on Windows, but higher up in `adbc_driver_manager.cc` there's an unconditional
```cpp
const char* env_path = std::getenv("ADBC_CONFIG_PATH");
```
Does that need to be changed?

ianmcook commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -273,14 +273,39 @@ AdbcStatusCode LoadDriverManifest(const std::filesystem::path& driver_manifest,
+    static const wchar_t* env_var = L"ADBC_CONFIG_PATH";
+#else
+    static const char* env_var = "ADBC_CONFIG_PATH";
+#endif
```
```suggestion
#endif  // _WIN32
```
